llvm/lib/Transforms/Utils/SSAUpdaterBulk.cpp

```diff
@@ -143,34 +136,66 @@ void SSAUpdaterBulk::RewriteAllUses(DominatorTree *DT,
     SmallVector<BasicBlock *, 32> IDFBlocks;
     SmallPtrSet<BasicBlock *, 32> LiveInBlocks;
     ComputeLiveInBlocks(UsingBlocks, DefBlocks, LiveInBlocks, PredCache);
-    IDF.resetLiveInBlocks();
     IDF.setLiveInBlocks(LiveInBlocks);
     IDF.calculate(IDFBlocks);
 
+    // Reserve map large enough to reduce growth.
+    BBInfos.init(LiveInBlocks.size() + DefBlocks.size());
+
+    for (auto [BB, V] : R.Defines)
+      BBInfos[BB].LiveOutValue = V;
+
     // We've computed IDF, now insert new phi-nodes there.
-    SmallVector<PHINode *, 4> InsertedPHIsForVar;
     for (auto *FrontierBB : IDFBlocks) {
       IRBuilder<> B(FrontierBB, FrontierBB->begin());
       PHINode *PN = B.CreatePHI(R.Ty, 0, R.Name);
-      R.Defines[FrontierBB] = PN;
-      InsertedPHIsForVar.push_back(PN);
+      BBInfos[FrontierBB].LiveInValue = PN;
       if (InsertedPHIs)
         InsertedPHIs->push_back(PN);
     }
 
+    // IsLiveOut indicates whether we are computing live-out values (true) or
+    // live-in values (false).
+    std::function<Value *(BasicBlock *, bool)> computeValue =
+        [&](BasicBlock *BB, bool IsLiveOut) -> Value * {
+      auto &BBInfo = BBInfos[BB];
+
+      if (IsLiveOut && BBInfo.LiveOutValue)
+        return BBInfo.LiveOutValue;
+
+      if (BBInfo.LiveInValue)
+        return BBInfo.LiveInValue;
+
+      Value *V = DT->isReachableFromEntry(BB) && PredCache.get(BB).size()
+                     ? computeValue(DT->getNode(BB)->getIDom()->getBlock(),
+                                    /* IsLiveOut = */ true)
+                     : UndefValue::get(R.Ty);
```

Comment:
Can this be poison

Reply:
SSAUpdater returns poison in the case, but I'm preserving the current behavior. I believe it's better to make this poison with a separate patch.